Add Order & Trade Models, restructure & rename order types

sangaman opened the pull request:
This PR closes #621 and resolves the first part of #437.

I set out to restructure the database models ahead of persisting trades to the database as we'd discussed, it wound up being a bit more complicated than anticipated (and this PR bigger than expected) but I tried to separate every conceptually distinct change into its own commit. Details are in the body of each commit message but I'll go over the important parts here:

- `initialQuantity` on stamped orders: Currently all `quantity` usage in `xud` refers only to the currently available quantity for an order, but we had nothing to track the initial quantity of an order before being matched, canceled, etc... I added this new property to stamped orders because I expect that it will be useful info when saving orders to the database. `initialQuantity` is set when we stamp an own order or when we receive a peer order.

- Renaming order types: Previously we'd discussed renaming the order types since the `StampedOrder`/`StampedOwnOrder`/`StampedPeerOrder` tyeps are ubiquitous whereas the plain `Order` type was unused. I renamed the three above to simply `Order`/`OwnOrder`/`PeerOrder` and renamed the previous `Order` type to `LimitOrder` which is just the existing `MarketOrder` type . The `LimitOrder` and `MarketOrder` types are just base types that aren't in use aside from inheritance. I also tried to use the `Pick` operator wherever possible so that types properly inherit properties (along with the comments) from base types. Functionally this doesn't change anything for now, but I did remove the `hold` property from the `OrderPacketBody` (since this is strictly local data that is not relevant to peers, in addition to being privileged info) which will make a difference when we start checking that packet bodies adhere to the defined structure.

- `Order` and `Trade` models. The `Order` model essentially mimics the info we have for `Order` in our proto definition - it has all the properties of the `Order` type (previously `StampedOrder`) minus `quantity` and `hold` since these are mutable properties that can be changed. The `Trade` model refers to a maker order, a taker order, and a quantity. Let me know if you think I've missed any properties, although we can always add them later if necessary.

- Since the `SwapDeal` model no longer has order-specific information, I added logic to check if a swapped order exists in the database and create it if it does not.

- Finally I figured some tests of the database models were in order. I only had time to go through basic "happy path" cases but I figure this would be a good suite of tests to expand in the future.

erkarl commented:
Running this branch locally and trying to execute a trade between XUD1 and XUD2 results in my nodes crashing.

CLI:
```
➜  xud git:(db/orders-table) ✗ xu2cli connect 033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b@localhost:8885
success
➜  xud git:(db/orders-table) ✗ xu1cli sell 0.00000001 LTC/BTC 120
{
  "internalMatchesList": [],
  "swapResultsList": [],
  "remainingOrder": {
    "price": 120,
    "quantity": 1e-8,
    "pairId": "LTC/BTC",
    "id": "68f51b81-dd0c-11e8-b609-e1af1530275c",
    "peerPubKey": "",
    "localId": "68f51b80-dd0c-11e8-b609-e1af1530275c",
    "createdAt": 1540990600504,
    "side": 1,
    "isOwnOrder": true
  }
}
➜  xud git:(db/orders-table) ✗ xu2cli buy 0.00000001 LTC/BTC 120
{
  "internalMatchesList": [],
  "swapResultsList": [
    {
      "orderId": "68f51b81-dd0c-11e8-b609-e1af1530275c",
      "localId": "709e9ff0-dd0c-11e8-a85a-f307a7a64233",
      "pairId": "LTC/BTC",
      "quantity": 1e-8,
      "rHash": "f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769",
      "amountReceived": 1,
      "amountSent": 120,
      "peerPubKey": "033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b",
      "role": 0
    }
  ]
}
```

XUD1:
```
[1] 2018-10-31 14:56:40 [ORDERBOOK] debug: order added: {"pairId":"LTC/BTC","price":120,"quantity":1e-8,"isBuy":false,"localId":"68f51b80-dd0c-11e8-b609-e1af1530275c","initialQuantity":1e-8,"id":"68f51b81-dd0c-11e8-b609-e1af1530275c","createdAt":1540990600504}
[1] 2018-10-31 14:56:53 [P2P] debug: received swapRequest from 031dcbcbd9d779bcb89095806cb10fd5f5f26d426ffcf037e6a4a338796c0b1b36: {"takerCurrency":"LTC","makerCurrency":"BTC","takerAmount":1,"makerAmount":120,"takerCltvDelta":576,"r_hash":"f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769","orderId":"68f51b81-dd0c-11e8-b609-e1af1530275c","pairId":"LTC/BTC","proposedQuantity":1e-8}
[1] 2018-10-31 14:56:53 [SWAPS] debug: New deal: {"takerCurrency":"LTC","makerCurrency":"BTC","takerAmount":1,"makerAmount":120,"takerCltvDelta":576,"r_hash":"f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769","orderId":"68f51b81-dd0c-11e8-b609-e1af1530275c","pairId":"LTC/BTC","proposedQuantity":1e-8,"takerPubKey":"03bc1d5f4e87b74282050ae8e2c0bf06eebe4901013c255198e893847a9271a196","peerPubKey":"031dcbcbd9d779bcb89095806cb10fd5f5f26d426ffcf037e6a4a338796c0b1b36","price":120,"localId":"68f51b80-dd0c-11e8-b609-e1af1530275c","quantity":1e-8,"phase":0,"state":0,"role":1,"createTime":1540990613362}
[1] 2018-10-31 14:56:53 [SWAPS] debug: got 1 routes to destination: 819070,,1,873982001709908000,4911373,1,,819070,1000,,1000
[1] 2018-10-31 14:56:53 [SWAPS] debug: got block height of 818494
[1] 2018-10-31 14:56:53 [SWAPS] debug: total timelock of route = 576makerCltvDelta = 288
[1] 2018-10-31 14:56:53 [SWAPS] debug: Sending swap response to peer
[1] 2018-10-31 14:56:53 [SWAPS] info: ResolveHash starting with hash: f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769
[1] 2018-10-31 14:56:53 [SWAPS] debug: Executing maker code
[1] 2018-10-31 14:56:53 [P2P] info: Peer 031dcbcbd9d779bcb89095806cb10fd5f5f26d426ffcf037e6a4a338796c0b1b36 socket closed
[1] 2018-10-31 14:56:53 [SWAPS] debug: Amount received for preImage 37b590a06dfa65b2dae778f0a891edc0410fea4916dd405329b798198631e3b3
[1] { message: 'order with id 031dcbcbd9d779bcb89095806cb10fd5f5f26d426ffcf037e6a4a338796c0b1b36/68f51b81-dd0c-11e8-b609-e1af1530275c could not be found',
[1]   code: '3.3' }
[1]
[1] /Users/ar/Projects/crypto/xud/dist/bootstrap.js:13
[1]         throw err;
[1]         ^
[1] [object Object]
[1] [nodemon] app crashed - waiting for file changes before starting...
```

XUD2:
```
[1] 2018-10-31 14:56:40 [P2P] verbose: received order from 033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b: {"pairId":"LTC/BTC","price":120,"quantity":1e-8,"isBuy":false,"initialQuantity":1e-8,"id":"68f51b81-dd0c-11e8-b609-e1af1530275c"}
[1] 2018-10-31 14:56:40 [ORDERBOOK] debug: order added: {"pairId":"LTC/BTC","price":120,"quantity":1e-8,"isBuy":false,"initialQuantity":1e-8,"id":"68f51b81-dd0c-11e8-b609-e1af1530275c","peerPubKey":"033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b","createdAt":1540990600505}
[1] 2018-10-31 14:56:53 [SWAPS] debug: New deal: {"takerCurrency":"LTC","makerCurrency":"BTC","takerAmount":1,"makerAmount":120,"takerCltvDelta":576,"r_hash":"f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769","orderId":"68f51b81-dd0c-11e8-b609-e1af1530275c","pairId":"LTC/BTC","proposedQuantity":1e-8,"peerPubKey":"033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b","localId":"709e9ff0-dd0c-11e8-a85a-f307a7a64233","price":120,"phase":0,"state":0,"r_preimage":"37b590a06dfa65b2dae778f0a891edc0410fea4916dd405329b798198631e3b3","role":0,"createTime":1540990613360}
[1] 2018-10-31 14:56:53 [SWAPS] debug: Requesting deal: {"takerCurrency":"LTC","makerCurrency":"BTC","takerAmount":1,"makerAmount":120,"takerCltvDelta":576,"r_hash":"f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769","orderId":"68f51b81-dd0c-11e8-b609-e1af1530275c","pairId":"LTC/BTC","proposedQuantity":1e-8,"peerPubKey":"033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b","localId":"709e9ff0-dd0c-11e8-a85a-f307a7a64233","price":120,"phase":0,"state":0,"r_preimage":"37b590a06dfa65b2dae778f0a891edc0410fea4916dd405329b798198631e3b3","role":0,"createTime":1540990613360}
[1] 2018-10-31 14:56:53 [P2P] debug: received swapResponse from 033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b: {"makerCltvDelta":288,"r_hash":"f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769","quantity":1e-8}
[1] 2018-10-31 14:56:53 [SWAPS] info: ResolveHash starting with hash: f66c70a1eb37ea5135678063e56c2d6a15760ec4eae68b63852d72d4bedcd769
[1] 2018-10-31 14:56:53 [SWAPS] debug: Executing taker code
[1] 2018-10-31 14:56:53 [SWAPS] debug: Amount received for preImage 37b590a06dfa65b2dae778f0a891edc0410fea4916dd405329b798198631e3b3
[1] { message: 'order with id 033fc9ee7befa3232e85eb84d7108000b671a91282b37dac289154a3b31e5f9f0b/68f51b81-dd0c-11e8-b609-e1af1530275c could not be found',
[1]   code: '3.3' }
[1]
[1] /Users/ar/Projects/crypto/xud/dist/bootstrap.js:13
[1]         throw err;
[1]         ^
[1] [object Object]
[1] [nodemon] app crashed - waiting for file changes before starting...
```

sangaman commented on lib/db/models/Currency.ts:
```diff
@@ -16,5 +16,18 @@ export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes)
+    models.Currency.hasMany(models.Pair, {
+      as: 'quoteCurrencies',
+      foreignKey: 'quoteCurrency',
+      constraints: true,
```
Oh that's odd, I thought I'd read that `constraints` was false by default. Although I think it doesn't hurt to be explicit - for example it turned out I needed to be explicit for `foreignKey` (see https://github.com/sequelize/sequelize/issues/9328) even though the defaults *should* have worked according to the docs.

I'll go and set `constraints` to false explicitly where it's not set to true, basically where the constraints would cause a problem right now are with the `Pair` model - when we remove a pair we remove the row from the database entirely which would violate the constraints for any existing orders/swaps for that pair. I was thinking maybe for a future PR we could have an `active` flag for Pair to indicate whether it is currently supported, but still leave a row in the table.

sangaman commented on lib/orderbook/OrderBookRepository.ts:
```diff
@@ -30,6 +31,15 @@ class OrderbookRepository {
     return this.models.Pair.bulkCreate(<db.PairAttributes[]>pairs);
   }
+
+  public addOrderIfNotExists = async (order: db.OrderFactory) => {
```
Yeah that's the path I started down but it was pulling back the entire instance which I didn't need.

sangaman commented on lib/db/models/Order.ts:
```diff
@@ -0,0 +1,48 @@
+export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes) => {
+  const attributes: db.SequelizeAttributes<db.OrderAttributes> = {
+    id: { type: DataTypes.STRING, primaryKey: true, allowNull: false },
+    peerNodeId: { type: DataTypes.STRING },
```
Yeah I was on the fence here as well. I might drop "peer" from the name and just add a comment to the `SwapDealAttributes` type to clarify what this property is for.

sangaman commented:
Thanks for testing, I'm investigating to see what the cause of the crash could have been.

sangaman commented:
OK, the reason for the crash was because of the new `persistOrderById` method - it was always being passed a `peerPubKey` because `SwapDeal` and `SwapResult` always have a value for `peerPubKey` - it's just the pub key of the peer that we're swapping with. This was the offending line:

```typescript
const order = peerPubKey ? this.getPeerOrder(orderId, pairId, peerPubKey) : this.getOwnOrder(orderId, pairId);
```

It would always try to get a peer order, even when the `orderId` refers to an own order. To fix it I made the logic check whether our role is maker or taker - `orderId` always refers to the maker order - to determine whether we need to get a peer order or own order and renamed the method to `persistSwapOrder`.

I also addressed the other comments, renaming `peerNodeId` to `nodeId` on the database level and while I was at it I noticed that it was a `STRING` when it should be an `INTEGER` to match the `id` of the `nodes` table.

sangaman commented on lib/db/models/Order.ts:
```diff
@@ -0,0 +1,48 @@
+  const attributes: db.SequelizeAttributes<db.OrderAttributes> = {
+    id: { type: DataTypes.STRING, primaryKey: true, allowNull: false },
+    peerNodeId: { type: DataTypes.STRING },
+    localId: { type: DataTypes.STRING },
```
Right I will explicitly specify it. Yes it's for both own and peer orders, so these are nullable. I can switch `peerNodeId` back, it's still a close call for me but I was figuring that we're not really using the "peer" terminology elsewhere in the database. The local node does not exist in the `nodes` table, so it only makes sense for `nodeId` to be set for remote orders. The attribute is commented within xud to try to make it clear. Let me know what you think.

moshababo replied:
Sounds fine to me.

sangaman commented on lib/db/models/Order.ts:
```diff
@@ -0,0 +1,48 @@
+    localId: { type: DataTypes.STRING },
+    initialQuantity: { type: DataTypes.DECIMAL(8), allowNull: true },
+    pairId: { type: DataTypes.STRING, allowNull: false },
+    price: { type: DataTypes.DECIMAL(8), allowNull: false },
```
Yeah good point, I believe `DECIMAL` types can handle that but I also don't think it's very graceful to store zero price and max number prices in the database. I think it might be nicest to make `price` nullable and leave it as null for market orders - as effectively they don't have a price. We have the `isBuy` property to indicate the direction of the order, so I don't think the price provides any meaningful information for market orders. Does that make sense?

sangaman replied:
I can handle that change in a future PR, for the time being we aren't saving any market orders to the database.

sangaman commented on lib/db/models/Trade.ts:
```diff
@@ -0,0 +1,33 @@
+    makerOrderId: { type: DataTypes.STRING, primaryKey: true },
+    takerOrderId: { type: DataTypes.STRING, primaryKey: true },
+    quantity: { type: DataTypes.DECIMAL(8), allowNull: false },
+  };
```
It's supposed to be for all trades/matches, whether internal or remote/swaps. `SwapDeal` currently tracks failed swaps, and a lot of other data we don't need if we're only interested in actual order executions - however they may have happened. I figure this table gives us exactly that. It holds very little information as you can see, just references to the orders and the quantity executed.

sangaman commented on lib/orderbook/OrderBookRepository.ts:
```diff
@@ -30,6 +31,15 @@ class OrderbookRepository {
     return this.models.Pair.bulkCreate(<db.PairAttributes[]>pairs);
   }
+
+  public addOrderIfNotExists = async (order: db.OrderFactory) => {
```
Yeah, I could not figure out a way to do this without resorting to raw SQL queries, and frankly I'm not sure how much more performant it is to do a `INSERT INTO... SELECT` vs this, but either way it's not a big deal because it's embedded like you said and it doesn't block the node process.

sangaman commented on lib/orderbook/OrderBook.ts:
```diff
@@ -87,16 +87,24 @@ class OrderBook extends EventEmitter {
     if (this.swaps) {
-      this.swaps.on('swap.paid', (swapResult) => {
+      this.swaps.on('swap.paid', async (swapResult) => {
+        await this.persistOrderById(swapResult);
```
This was buggy per @erkarl's crash report so I've changed it some already if you want to take another look, but to answer your questions:

1. Yes this was indeed the bug and was fixed I believe.

2. Basically, if it's our own maker order, we won't even know about the peer order. All we know is a peer wanted to swap with us. If it's a peer maker order, I figure we will want to persist our own taker order, but I'm thinking that should happen as part of saving the trade to the db. For this PR I was mainly just focusing on db schema/structure and not breaking any existing functionality.

3. Yes we do, but that would be same as 2 above where that should happen when we persist the trade, which we're not doing yet, and can be handled in a future PR.

4. No particular reason, I don't think the order matters here though.

sangaman replied:
Actually for 4, the order does matter, because we want to persist the order before we remove it. After we've removed it, we won't have the information any longer.

moshababo replied:
I continued the discussion in the main thread, since you already were commented about some of these things.

sangaman commented:
@moshababo Thanks for the review (and for independently catching that bug with the `persistOrderById` method). See what you think of my responses to your comments and I can make changes in a little while.

sangaman commented:
There was still an issue with the orders for swaps were being persisted, because some of the existing matching/swap execution logic actually removes the peer order immediately upon beginning a swap. This makes sense, because whether the swap succeeds or fails we're going to want to remove it from the book, but then when I was going to retrieve the order information to persist it, it was no longer there.

Instead, I'm persisting the order right when a deal is proposed or accepted. This follows similar logic, whether it fails or succeeds we will want a record of it in the database. This doesn't require any `getOwnOrder` or `getPeerOrder` lookup, so there's no possibility of trying to find an order that's not there.

moshababo commented:
You mentioned that there’s no particular reason for persisting the order just before the removal, in contrast to doing it earlier. 

I’d say that doing it before the removal is much more complicated, either because you don't have the order object, like previously, or just because there are many paths to cover, so that might lead to future bugs. Persisting the order in the beginning is simple and straightforward (you just need to hook to `addOwnOrder` / `addPeerOrder`).

The downsides: you need to flush the order state updates (currently just `quantity` updates I believe), and it will also save orders which ended up not being involved in a trade. 

But I’m not sure if that’s what you intent, because at the moment you are persisting an order which might not end up on a trade (`OrderBook.ts` line 304). If so, I think we can consider persisting when adding, and not when removing.

Mind that we also have the bulk peer orders removal (upon disconnection), which are a bunch of orders that were not involved in a trade. 

We don’t need to address this in this PR, since it’s for the DB schema. I commented about the main schema design (`swapdeals`/`trades` tables) above.

moshababo commented on lib/db/models/SwapDeal.ts:
```diff
@@ -9,16 +9,14 @@ export default (sequelize: Sequelize.Sequelize, DataTypes: Sequelize.DataTypes)
-    peerPubKey: { type: DataTypes.STRING, allowNull: false },
+    peerNodeId: { type: DataTypes.STRING, allowNull: false },
     orderId: { type: DataTypes.STRING, allowNull: false },
     localId: { type: DataTypes.STRING, allowNull: false },
```
We can leave this out for now, i'll think about it some more when doing the swaps unit tests.

sangaman commented:
> I’d say that doing it before the removal is much more complicated, either because you don't have the order object, like previously, or just because there are many paths to cover, so that might lead to future bugs. Persisting the order in the beginning is simple and straightforward (you just need to hook to addOwnOrder / addPeerOrder).
> 
> The downsides: you need to flush the order state updates (currently just quantity updates I believe), and it will also save orders which ended up not being involved in a trade.

Yes I think you're right and with the current approach I'm persisting the order when the swap begins. It would be possible (and I'd agree probably most straightforward) to save the order to the database immediately when we add it like you say, but then we'd be storing every order that ever goes through `xud` like you say. Maybe that's something we'd want, we can discuss this possibility, but I'd lean against it because it sounds like potentially a lot of data with minimal value given we never executed it or even attempted to execute it. If we do decide that approach I think it'd be quite easy to switch in the future though.

> But I’m not sure if that’s what you intent, because at the moment you are persisting an order which might not end up on a trade (OrderBook.ts line 304). If so, I think we can consider persisting when adding, and not when removing.

Line 304 persists an order that's going to be attempted for a swap - it might not succeed and wind up in a trade but since we save swap deals that end in failure, this is essentially preserving the previous behavior except putting the order-related info in a separate table. This will let us look up an order and get all the swaps for that order, whether they succeeded or failed.

moshababo commented:
Ok, so I understand that we don’t want to save orders which are not involved in a swap attempt or an internal match. It basically excludes orders of a peer which got disconnected, and all pending orders inside `xud`, which will be lost in case of a shutdown. 

If so, it is better to save them upon removal, but we need to make sure that the code we write for it is maintainable. My suggestion is to combine it with our orders removal events, which need to cover all the cases anyway. The basic idea is to have a distinct event for each case (the current situation, except for `peerOrder.invalidation`, which serves a double purpose), and to create a small wrapper method for each one to be called for emitting it, in addition to persisting the order. 

In regards to the `trades` table persisting successful swaps - I don’t see a problem with using a database view in order to retrieve the complete trades list (successful swaps + internal matches), but it isn’t a big deal. 

So in the overall, good work here, I think it's ready to be merged.

erkarl commented:
@sangaman good to merge?
